sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -724,6 +724,16 @@ object LimitPushDown extends Rule[LogicalPlan] {
       Limit(le, Project(a.aggregateExpressions, LocalLimit(le, a.child)))
     case Limit(le @ IntegerLiteral(1), p @ Project(_, a: Aggregate)) if a.groupOnly =>
       Limit(le, p.copy(child = Project(a.aggregateExpressions, LocalLimit(le, a.child))))
+    // Only push down when the limit number is less than or equal to 5000,
+    // which is the default_limit value of Hue.
+    case Limit(le @ IntegerLiteral(limit), a @ Aggregate(_, _, child))
+      if a.groupOnly && child.maxRowsPerPartition.forall(_ > limit) &&
+        limit <= math.min(conf.topKSortFallbackThreshold, 5000) =>
```

Comment:
We do not push down all cases because number of limit rows need to do an extra `HashAggregate`. The maximum threshold is 5000 to benefit most SQL editors:
https://github.com/cloudera/hue/blob/release-4.10.0/desktop/conf.dist/hue.ini#L948-L949